pkg/auth/providers/ldap/ldap_client.go

```diff
@@ -0,0 +1,504 @@
+	serviceAccountUsername := ldap.GetUserExternalID(config.ServiceAccountUsername, "")
+	err = lConn.Bind(serviceAccountUsername, config.ServiceAccountPassword)
+
+	if err != nil {
+		if ldapv2.IsErrorWithCode(err, ldapv2.LDAPResultInvalidCredentials) && config.Enabled {
+			var kind string
+			if strings.EqualFold("user", userType) {
+				kind = "user"
+			} else if strings.EqualFold("group", userType) {
+				kind = "group"
+			}
+			principal := &v3.Principal{
+				ObjectMeta:    metav1.ObjectMeta{Name: scope + "://" + distinguishedName},
+				DisplayName:   distinguishedName,
+				LoginName:     distinguishedName,
+				PrincipalType: kind,
+			}
+
+			return principal, nil
+		}
+		return nil, fmt.Errorf("Error in ldap bind: %v", err)
+	}
+
+	if strings.EqualFold("user", userType) {
+		search = ldapv2.NewSearchRequest(distinguishedName,
+			ldapv2.ScopeBaseObject, ldapv2.NeverDerefAliases, 0, 0, false,
+			filter,
+			p.getUserSearchAttributes(config), nil)
+	} else {
+		search = ldapv2.NewSearchRequest(distinguishedName,
+			ldapv2.ScopeBaseObject, ldapv2.NeverDerefAliases, 0, 0, false,
+			filter,
+			p.getGroupSearchAttributes(config), nil)
+	}
+
+	result, err := lConn.Search(search)
+	if err != nil {
+		if ldapErr, ok := err.(*ldapv2.Error); ok && ldapErr.ResultCode == 32 {
+			return nil, httperror.NewAPIError(httperror.NotFound, fmt.Sprintf("%v not found", distinguishedName))
+		}
+		return nil, httperror.WrapAPIError(errors.Wrapf(err, "server returned error for search %v %v: %v", search.BaseDN, filter, err), httperror.ServerError, "Internal server error")
+	}
+
+	if len(result.Entries) < 1 {
+		return nil, fmt.Errorf("No identities can be retrieved")
+	} else if len(result.Entries) > 1 {
+		return nil, fmt.Errorf("More than one result found")
+	}
+
+	entry := result.Entries[0]
+	entryAttributes := entry.Attributes
+
+	if !p.permissionCheck(entry.Attributes, config) {
+		return nil, fmt.Errorf("Permission denied")
+	}
+
+	principal, err := p.attributesToPrincipal(entryAttributes, distinguishedName, scope, config)
+	if err != nil {
+		return nil, err
+	}
+	if principal == nil {
+		return nil, fmt.Errorf("Principal not returned for LDAP")
+	}
+	return principal, nil
+}
+
+func (p *ldapProvider) attributesToPrincipal(attribs []*ldapv2.EntryAttribute, dnStr string, scope string, config *v3.LdapConfig) (*v3.Principal, error) {
+	var externalIDType, accountName, externalID, login string
+	var principal *v3.Principal
+	externalID = dnStr
+	externalIDType = scope
+
+	if ldap.IsType(attribs, config.UserObjectClass) {
+		for _, attr := range attribs {
+			if attr.Name == config.UserNameAttribute {
+				if len(attr.Values) != 0 {
+					accountName = attr.Values[0]
+				} else {
+					accountName = externalID
+				}
+			}
+			if attr.Name == config.UserLoginAttribute {
+				if len(attr.Values) > 0 && attr.Values[0] != "" {
+					login = attr.Values[0]
+				}
+			} else {
+				login = accountName
```

Comment:
This else block is wrong. It should be dropped. If you have 5 attributes in attribs and the firest one is the userLoginAttribute, you'll set login to the right value but on the next iteration of the for loop, this else block is going to cause you to set login to the second attributes value. It will eventually end up as whatever the last attribute is.

Reply:
Probably the correct thing to do is move the else block inside the `if attr.Name == config.UserLoginAttribute {`` if block

Reply:
But then, there is not garauntee that accountName has been set yet. The right thing to do is to check if login == "" after you exit the for loop and if it is, set it to accountName.

PS - I realize that this was a cary over bugy from active directory